opencog/cogserver/server/AgentRunnerBase.cc

```diff
@@ -0,0 +1,165 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program; if not, write to:
+ * Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <algorithm>
+#include <chrono>
+#include <opencog/cogserver/server/AgentRunnerBase.h>
+#include <opencog/cogserver/server/CogServer.h>
+#include <opencog/util/Logger.h>
+#include <opencog/util/platform.h>
+
+using namespace std;
+using namespace chrono;
+
+
+namespace opencog
+{
+
+struct equal_to_id: public std::binary_function<AgentPtr, const std::string&, bool>
+{
+        bool operator()(AgentPtr a, const std::string& cid) const
+        {
+            return (a->classinfo().id != cid);
+        }
+};
+
+AgentRunnerBase::AgentRunnerBase(std::string runner_name): name(runner_name),
+        cycleCount(1)
+{
+}
+
+AgentRunnerBase::~AgentRunnerBase()
+{
+}
+
+void AgentRunnerBase::setName(std::string new_name)
+{
+    name = new_name;
+}
+
+const std::string& AgentRunnerBase::getName() const
+{
+    return name;
+}
+
+unsigned long AgentRunnerBase::getCycleCount() const
+{
+    return cycleCount;
+}
+
+void AgentRunnerBase::addAgent(AgentPtr a)
+{
+    agents.push_back(a);
+}
+
+void AgentRunnerBase::removeAgent(AgentPtr a)
+{
+    AgentSeq::iterator ai = std::find(agents.begin(), agents.end(), a);
+    if (ai != agents.end()) {
+        agents.erase(ai);
+        cogserver().systemActivityTable().clearActivity(a);
+        a->stop();
+        logger().debug("[CogServer::%s] stopped agent \"%s\"", name.c_str(),
+            a->to_string().c_str());
+    }
+}
+
+void AgentRunnerBase::removeAllAgents(const std::string& id)
+{
+    // place agents with classinfo().id == id at the end of the container
+    AgentSeq::iterator last =
+        std::partition(agents.begin(), agents.end(),
+                       boost::bind(equal_to_id(), _1, id));
+
+    std::for_each(last, agents.end(), [] (AgentPtr &a) {
+        cogserver().systemActivityTable().clearActivity(a);
+        a->stop();
+    });
+
+    // remove those agents from the main container
+    agents.erase(last, agents.end());
+
+    logger().debug("[CogServer::%s] stopped all agents of type \"%s\"",
+        name.c_str(), id.c_str());
+}
+
+void AgentRunnerBase::removeAllAgents()
+{
+    for (auto &a: agents)
+    {
+        cogserver().systemActivityTable().clearActivity(a);
+        a->stop();
+    }
+    agents.clear();
+    logger().debug("[CogServer::%s] stopped all agents", name.c_str());
+}
+
+void AgentRunnerBase::runAgent(AgentPtr a)
+{
+    size_t mem_start = getMemUsage();
+    int atoms_start = atomspace().get_size();
+
+    logger().debug("[CogServer::%s] begin to run mind agent: %s, [cycle = %d]",
+                   name.c_str(), a->classinfo().id.c_str(), cycleCount);
+
+    auto timer_start = system_clock::now();
+
+    a->resetUtilizedHandleSets();
+    a->run();
```

Comment:
Both is possible. And, by default, agents are run like before: cooperative time-slicing in the main server thread. However, it is possible to run agents in dedicated threads too. These threads might be named, in which case it is possible to assign more than one agent to a single worker thread, and it'll run all of them cooperatively (however, it ignores the 'frequency' parameter: all of them run in each cycle). 

Also note that an agent's run() method is still supposed to run a single cycle of agent process, like before.